src/main/java/bio/terra/service/snapshotbuilder/utils/CriteriaQueryBuilder.java

```diff
@@ -0,0 +1,211 @@
+        SubQueryFilterVariable.Operator.IN,
+        new Query(
+            List.of(
+                new FieldVariable(
+                    new FieldPointer(occurrencePointer, PERSON_ID_FIELD_NAME), occurrenceVariable)),
+            List.of(occurrenceVariable, ancestorVariable),
+            new BooleanAndOrFilterVariable(
+                BooleanAndOrFilterVariable.LogicalOperator.OR,
+                List.of(
+                    new BinaryFilterVariable(
+                        new FieldVariable(
+                            new FieldPointer(occurrencePointer, occurrenceTable.idColumnName()),
+                            occurrenceVariable),
+                        BinaryFilterVariable.BinaryOperator.EQUALS,
+                        new Literal(domainCriteria.getId().intValue())),
+                    new BinaryFilterVariable(
+                        new FieldVariable(
+                            new FieldPointer(ancestorPointer, "ancestor_concept_id"),
+                            ancestorVariable),
+                        BinaryFilterVariable.BinaryOperator.EQUALS,
+                        new Literal(domainCriteria.getId().intValue()))))));
+  }
+
+  public FilterVariable generateFilterForCriteria(SnapshotBuilderCriteria criteria) {
+    return switch (criteria.getKind()) {
+      case LIST -> {
+        if (criteria instanceof SnapshotBuilderProgramDataListCriteria listCriteria) {
+          yield generateFilterForListCriteria(listCriteria);
+        }
+        throw new BadRequestException("Malformed list criteria");
+      }
+      case RANGE -> {
+        if (criteria instanceof SnapshotBuilderProgramDataRangeCriteria rangeCriteria) {
+          yield generateFilterForRangeCriteria(rangeCriteria);
+        }
+        throw new BadRequestException("Malformed range criteria");
+      }
+      case DOMAIN -> {
+        if (criteria instanceof SnapshotBuilderDomainCriteria domainCriteria) {
+          yield generateFilterForDomainCriteria(domainCriteria);
+        }
+        throw new BadRequestException("Malformed domain criteria");
+      }
+    };
+  }
+
+  public FilterVariable generateAndOrFilterForCriteriaGroup(
+      SnapshotBuilderCriteriaGroup criteriaGroup) {
+    return new BooleanAndOrFilterVariable(
+        criteriaGroup.isMeetAll() != null && criteriaGroup.isMeetAll()
+            ? BooleanAndOrFilterVariable.LogicalOperator.AND
+            : BooleanAndOrFilterVariable.LogicalOperator.OR,
+        criteriaGroup.getCriteria().stream().map(this::generateFilterForCriteria).toList());
+  }
+
+  public FilterVariable generateFilterForCriteriaGroup(SnapshotBuilderCriteriaGroup criteriaGroup) {
+    if (criteriaGroup.isMustMeet() != null && criteriaGroup.isMustMeet()) {
+      return generateAndOrFilterForCriteriaGroup(criteriaGroup);
+    } else {
+      return new NotFilterVariable(generateAndOrFilterForCriteriaGroup(criteriaGroup));
+    }
```

Comment:
You could refactor this so there's only one call to `generateAndOrFilterForCriteriaGroup()`
```suggestion
    FilterVariable groupFilter = generateAndOrFilterForCriteriaGroup(criteriaGroup);
    if (criteriaGroup.isMustMeet() == null || !criteriaGroup.isMustMeet()) {
      return new NotFilterVariable(groupFilter);
    }
    return groupFilter;
```